src/kern/npf_alg_pptp.c

```diff
@@ -0,0 +1,763 @@
+
+	/* allocate */
+	alg_arg = kmem_zalloc(sizeof(struct pptp_alg_arg), KM_SLEEP);
+	if (alg_arg == NULL)
+		return NULL;
+
+	/* mark all entries as empty */
+	for (i = 0; i < PPTP_MAX_GRE_PER_CLIENT; i++)
+		alg_arg->gre_cons[i].flags = PPTP_ALG_FL_FREE_ENTRY;
+
+	return alg_arg;
+}
+
+/*
+ *
+ */
+static struct pptp_gre_con *
+npfa_pptp_alg_arg_lookup(struct pptp_alg_arg *arg)
+{
+	int i;
+
+	for (i = 0; i < PPTP_MAX_GRE_PER_CLIENT; i++)
+		if (arg->gre_cons[i].flags & PPTP_ALG_FL_FREE_ENTRY)
+			return &arg->gre_cons[i];
+
+	return NULL;
+}
+
+/*
+ *
+ */
+static struct pptp_gre_con *
+npfa_pptp_alg_arg_lookup_with_server_call_id(struct pptp_alg_arg *arg,
+		  uint16_t server_call_id)
+{
+	int i;
+	struct pptp_gre_con *gre_con;
+
+	for (i = 0; i < PPTP_MAX_GRE_PER_CLIENT; i++) {
+		gre_con = &arg->gre_cons[i];
+
+		if ((gre_con->flags & PPTP_ALG_FL_FREE_ENTRY) == 0 &&
+				  gre_con->ctx.server_call_id == server_call_id)
+			return gre_con;
+	}
+
+	return NULL;
+}
+
+/*
+ *
+ */
+static struct pptp_gre_con *
+npfa_pptp_alg_arg_lookup_with_orig_client_call_id(struct pptp_alg_arg *arg,
+		  uint16_t client_call_id)
+{
+	int i;
+	struct pptp_gre_con *gre_con;
+
+	for (i = 0; i < PPTP_MAX_GRE_PER_CLIENT; i++) {
+		gre_con = &arg->gre_cons[i];
+
+		if ((gre_con->flags & PPTP_ALG_FL_FREE_ENTRY) == 0 &&
+				  gre_con->orig_client_call_id == client_call_id)
+			return gre_con;
+	}
+
+	return NULL;
+}
+
+/*
+ *
+ */
+static struct pptp_gre_con *
+npfa_pptp_alg_arg_lookup_with_client_call_id(struct pptp_alg_arg *arg,
+		  uint16_t client_call_id)
+{
+	int i;
+	struct pptp_gre_con *gre_con;
+
+	for (i = 0; i < PPTP_MAX_GRE_PER_CLIENT; i++) {
+		gre_con = &arg->gre_cons[i];
+
+		if ((gre_con->flags & PPTP_ALG_FL_FREE_ENTRY) == 0 &&
+				  gre_con->ctx.client_call_id == client_call_id)
+			return gre_con;
+	}
+
+	return NULL;
+}
+
+/*
+ * PPTP TCP control connection ALG translator.
+ * It rewrites Call ID in the Outgoing-Call-Request
+ * message and Peer Call ID in the Outgoing-Call-Reply message.
+ */
+static bool
+npfa_pptp_tcp_translate(npf_cache_t *npc, npf_nat_t *nt, bool forw)
+{
+	uint16_t cksum;
+	uint16_t old_call_id;
+	uint32_t tcp_hdr_size;
+	uint32_t ip;
+	nbuf_t *nbuf;
+	struct tcphdr *tcp;
+	struct pptp_msg_hdr *pptp;
+	struct pptp_outgoing_call_reply *pptp_call_reply;
+	struct pptp_alg_arg *alg_arg;
+	struct pptp_gre_con *gre_con;
+	npf_cache_t gre_npc;
+	npf_addr_t *o_addr;
+	in_port_t o_port;
+
+	/* only ipv4 is supported so far */
+	if (!(npf_iscached(npc, NPC_IP4) && npf_iscached(npc, NPC_TCP) &&
+			  (npc->npc_l4.tcp->th_dport == htons(PPTP_SERVER_PORT) ||
+			  npc->npc_l4.tcp->th_sport == htons(PPTP_SERVER_PORT))))
+		return false;
+
+	nbuf = npc->npc_nbuf;
+	tcp = npc->npc_l4.tcp;
+	tcp_hdr_size = tcp->th_off << 2;
+	nbuf_reset(nbuf);
+
+	pptp = nbuf_advance(nbuf, npc->npc_hlen + tcp_hdr_size,
+			  sizeof(struct pptp_msg_hdr));
+	if (pptp == NULL)
+		return false;
+
+	if (pptp->pptp_msg_type != htons(PPTP_CTRL_MSG) ||
```

Comment:
You need to use `nbuf_ensure_contig()` for the size of the header to ensure that it is in the contiguous buffer. Otherwise, accessing the header is not safe.

Reply:
nbuf_advance() does that for me

Reply:
I've added nbuf_ensure_contig() for PPTP_OUTGOING_CALL_REPLY case. It's the only case where there is a cast to a more specific (longer) PPTP structure.